libtransmission/rpc-server.cc

```diff
@@ -999,39 +1167,57 @@ tr_rpc_server::tr_rpc_server(tr_session* session_in, tr_variant* settings)
         tr_rpcSetAntiBruteForceThreshold(this, i);
     }
 
+    key = TR_KEY_rpc_socket_mode;
+
+    if (!tr_variantDictFindInt(settings, key, &i))
+    {
+        missing_settings_key(key);
+    }
+    else
+    {
+        tr_rpcSetRPCSocketMode(this, i);
+    }
+
     key = TR_KEY_rpc_bind_address;
 
     if (!tr_variantDictFindStrView(settings, key, &sv))
     {
         missing_settings_key(key);
-        address = tr_inaddr_any;
+        address = tr_rpc_inaddr_any;
     }
     else
     {
-        if (!tr_address_from_string(&address, std::string{ sv }.c_str()))
+        if (!tr_rpc_address_from_string(&address, sv))
         {
-            tr_logAddNamedError(MyName, _("%" TR_PRIsv " is not a valid address"), TR_PRIsv_ARG(sv));
-            address = tr_inaddr_any;
+            tr_logAddNamedError(MyName, _("%" TR_PRIsv " is not a valid address, falling back to 0.0.0.0."), TR_PRIsv_ARG(sv));
+            address = tr_rpc_inaddr_any;
         }
-        else if (address.type != TR_AF_INET && address.type != TR_AF_INET6)
+        else if (address.type != TR_RPC_AF_INET && address.type != TR_RPC_AF_INET6 && address.type != TR_RPC_AF_UNIX)
         {
             tr_logAddNamedError(
                 MyName,
-                _("%" TR_PRIsv " is not an IPv4 or IPv6 address. RPC listeners must be IPv4 or IPv6"),
+                _("%" TR_PRIsv
+                  " is not an IPv4 address, an IPv6 address, or a unix socket path. RPC listeners must be one of the previously mentioned types. Falling back to 0.0.0.0."),
                 TR_PRIsv_ARG(sv));
-            address = tr_inaddr_any;
+            address = tr_rpc_inaddr_any;
         }
     }
 
-    this->bindAddress = address;
+    if (address.type == TR_RPC_AF_UNIX)
+    {
+        this->isWhitelistEnabled = false;
+        this->isHostWhitelistEnabled = false;
+    }
+
+    this->bindAddress = (tr_rpc_address*)tr_memdup(&address, sizeof(tr_rpc_address));
 
     if (this->isEnabled)
     {
+
```

Comment:
Despite allowed by the formatter, we shouldn't be adding empty lines at the beginning of code blocks. In case someone's wondering, it's only allowed so that you can do so within the top-level `namespace` and `extern` blocks, where the body is not indented.